aya-bpf-macros/src/lib.rs

```diff
@@ -128,6 +128,35 @@ pub fn sk_msg(attrs: TokenStream, item: TokenStream) -> TokenStream {
     }
 }
 
+/// Marks a function as an eBPF XDP program that can be attached to a network interface.
+///
+/// On some NIC drivers, XDP probes are compatible with jumbo frames through the use of
```

Comment:
This comment is confusing. You get jumbo frames even without frags, they're just
always linearised

Reply:
uuuh no, unless I'm mistaken, you cannot attach an xdp program not marked frags on an interface where the mtu is greater than page size - headroom. I've only managed to do it on mellanox connectx-6 hardware with a vendor-specific flag turned on with ethtool and the frags option. afaik only mlx5 and i40e have support for frags (and i40e was non-functional last time I tried)

Reply:
AFAIK you can attach a non-frags program to a jumbo frame-enabled interface just fine - where jumbo frames mean MTU set to anything with MTU > 1500. Without frags support, you only get up to a page of data.
I'll double-check this next week though.

Reply:
you can as long as the jumbo frame + xdp headroom fits in a page (4096 bytes):

ebpf:
```rust
#![no_std]
#![no_main]

use aya_bpf::{bindings::xdp_action::XDP_PASS, macros::xdp, programs::XdpContext};

#[xdp] fn xdp_no_frags(_ctx: XdpContext) -> u32 { XDP_PASS }
#[xdp] fn xdp_with_frags(_ctx: XdpContext) -> u32 { XDP_PASS }
```

then the userspace either one or the other depending on the `--frags` flag:

```rust
use std::convert::TryInto;

use aya::{Bpf, include_bytes_aligned, programs::{Xdp, XdpFlags}};
use anyhow::Context;
use structopt::StructOpt;

fn main() {
    if let Err(e) = try_main() { eprintln!("error: {:#}", e); }
}

#[derive(Debug, StructOpt)]
struct Opt {
    #[structopt(short, long, default_value = "eth0")] iface: String,
    #[structopt(short, long)] frags: bool,
}

fn try_main() -> Result<(), anyhow::Error> {
    let opt = Opt::from_args();
    let mut bpf = Bpf::load(include_bytes_aligned!(
        "../../target/bpfel-unknown-none/release/aya-test"
    ))?;

    let name = if opt.frags { "xdp_with_frags" } else { "xdp_no_frags" };
    let program: &mut Xdp = bpf.program_mut(name).unwrap().try_into()?;
    program.load()?;
    program.attach(&opt.iface, XdpFlags::DRV_MODE).context("failed to attach XDP program in driver mode")?;

    println!("loaded xdp program {name}");

    Ok(())
}
```

testing on a connectx-6 dx nic:

```
root@cx6:~# ip l show dev enp23s0np0
4: enp23s0np0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN mode DEFAULT group default qlen 1000
    link/ether b8:ce:f6:bc:cb:4c brd ff:ff:ff:ff:ff:ff
root@cx6:~# ./aya-test -i enp23s0np0
loaded xdp program xdp_no_frags
root@cx6:~# ip l set dev enp23s0np0 mtu 3000 # still fits in a page
root@cx6:~# ./aya-test -i enp23s0np0
loaded xdp program xdp_no_frags
root@cx6:~# ip l set dev enp23s0np0 mtu 4000
root@cx6:~# ./aya-test -i enp23s0np0
error: failed to attach XDP program in driver mode: `bpf_link_create` failed: Invalid argument (os error 22)
root@cx6:~# ip l set dev enp23s0np0 mtu 9000
root@cx6:~# ./aya-test -i enp23s0np0
error: failed to attach XDP program in driver mode: `bpf_link_create` failed: Invalid argument (os error 22)
root@cx6:~# ./aya-test -i enp23s0np0 --frags
loaded xdp program xdp_with_frags
```

so yes, with jumbo being "anything above > 1500 bytes", frags are not _always_ required. however, this is the strict definition of jumbo, while the usual definition is mtu 9k, in which case frags are _definitely_ always required. this comes from the fact that, for performance reasons, xdp contexts including payload and head/tailroom is required to fit in a single page. packets thus get fragmented with pointers to other frags if it does not fit in a single page. the kernel requires this opt-in as the packet is not linear anymore, and the program needs some bpf helpers to access the payload further.

Reply:
> I don't think so? Maybe it depends on the attach mode? On aws I've definitely worked with 9001 bytes large frames with regular (no frags) xdp programs

ha missed this part. yes skb mode does not have this restriction, and will linearize the data. driver mode however is another beast, and to get the most of your hardware is required. on virtualized stuff it does not really matter, because unless the host does nic vf + pcie passthrough, it's quite likely it already generated a sk_buff for the packet, and passed it directly to the gest because virtio. (dunno about ena tho)